plpgsql: fix handling of annotations for DO blocks

yuzefovich opened the pull request:
**plpgsql: fix handling of annotations for DO blocks**

This commit should fix for good annotations handling for DO blocks that
was recently attempted to be fixed in 3b8ab38a67a7792200bf033a850513359fb527a8.
(That commit is partially reversed too).

The general issue is that in the PLpgSQL parser we use an ephemeral SQL
parser instance to process SQL statements. However, when processing the
DO block stmts we need to preserve the annotations state between each
stmt. This is now achieved by extending the SQL parser infrastructure to
give an option to override the initial annotations index as well as keep
on reusing the counter across multiple stmts. This allows us to
correctly count the number of annotations that we need for the whole DO
block during the initial parsing, and then we use the allocated
annotations container to set the items in the optbuild.

Fixes: #151848.

Release note (bug fix): Previously, CockroachDB node could crash when
executing DO stmts when they contain user-defined types (possibly
non-existing) in non-default configuration (additional logging like the
one controlled via `sql.log.all_statements.enabled` cluster setting needed
to be enabled). This bug was introduced in 25.1 release and is now fixed.

**tree: avoid index-of-bounds panics in GetAnnotation in production**

We've recently seen two panics when trying to access the Annotations
container and hitting the out-of-bounds condition. The annotations are
used to provide additional information about unresolved object names, so
if we simply return `nil` on an invalid access, we'll keep on using the
unresolved object name, which seems like a better option than process
crash. In test builds we'll keep on crashing.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/151849)
<!-- Reviewable:end -->

yuzefovich committed:
plpgsql: fix handling of annotations for DO blocks

This commit should fix for good annotations handling for DO blocks that
was recently attempted to be fixed in 3b8ab38a67a7792200bf033a850513359fb527a8.
(That commit is partially reversed too).

The general issue is that in the PLpgSQL parser we use an ephemeral SQL
parser instance to process SQL statements. However, when processing the
DO block stmts we need to preserve the annotations state between each
stmt. This is now achieved by extending the SQL parser infrastructure to
give an option to override the initial annotations index as well as keep
on reusing the counter across multiple stmts. This allows us to
correctly count the number of annotations that we need for the whole DO
block during the initial parsing, and then we use the allocated
annotations container to set the items in the optbuild.

Release note (bug fix): Previously, CockroachDB node could crash when
executing DO stmts when they contain user-defined types (possibly
non-existing) in non-default configuration (additional logging like the
one controlled via `sql.log.all_statements.enabled` cluster setting needed
to be enabled). This bug was introduced in 25.1 release and is now fixed.

yuzefovich committed:
tree: avoid index-of-bounds panics in GetAnnotation in production

We've recently seen two panics when trying to access the Annotations
container and hitting the out-of-bounds condition. The annotations are
used to provide additional information about unresolved object names, so
if we simply return `nil` on an invalid access, we'll keep on using the
unresolved object name, which seems like a better option than process
crash. In test builds we'll keep on crashing.

Release note: None

michae2 reviewed (changes requested):
Makes sense, I just have some questions since I'm a little unfamiliar with this code.

> @michae2 reviewed 11 of 11 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-:-OY8dMwN7B85cXDPZY0c:b5b1oll)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/plpgsql/parser/parse.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8ZU5g0aGXwuY5wf9g:-OY8ZU5g0aGXwuY5wf9h:b7qvcb) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/plpgsql/parser/parse.go#L137)):*
> ```Go
> // Parse parses a sql statement string and returns a list of Statements.
> // TODO(yuzefovich): audit the callers to ensure that NumAnnotations is
> // consulted wherever applicable.
> ```

What will happen to other callers of Parse? It seems like we will use unique annotation indexes, and will set NumAnnotations, but then not actually save the annotations anywhere? Will that interfere with formatting of a CREATE FUNCTION statement for example?
___
*[`pkg/sql/logictest/testdata/logic_test/do` line 305 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY88mei1Hf6aNH23loU:-OY88mei1Hf6aNH23loV:bxh2pu0) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/logictest/testdata/logic_test/do#L305)):*
> ```
> BEGIN
> END;
> $$;
> ```

Is there a way to test DO blocks that call other DO blocks?
___
*[`pkg/sql/plpgsql/parser/lexer.go` line 188 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8W0lYBSYe4bOfRAh3:-OY8W0lYBSYe4bOfRAh4:b-5ptmco) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/plpgsql/parser/lexer.go#L188)):*
> ```Go
> 		return nil, pgerror.New(pgcode.Syntax, "INTO used with a command that cannot return data")
> 	}
> 	l.numAnnotations = sqlStmt.NumAnnotations
> ```

If this has to be done after every call to ParseOneWithOptions, would it make sense to wrap ParseOneWithOptions in a helper function?
___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 856 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8Z9yM9OT8ToAiEYSe:-OY8Z9yM9OT8ToAiEYSf:b-2eovha) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/sem/plpgsqltree/statements.go#L856)):*
> ```Go
> 	StatementImpl
> 	SqlStmt     tree.Statement
> 	Annotations *tree.Annotations
> ```

Should this be removed?
___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 984 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8Z4wrCxchXZFTVBXX:-OY8Z4wrCxchXZFTVBXY:b-2eovha) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/sem/plpgsqltree/statements.go#L984)):*
> ```Go
> 	Strict      bool // INTO STRICT flag
> 	Target      []Variable
> 	Annotations *tree.Annotations
> ```

Should this be removed?
___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 1113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8XSoq5nMNMLhtV4Ij:-OY8XSoq5nMNMLhtV4Ik:bqup3sy) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/sem/plpgsqltree/statements.go#L1113)):*
> ```Go
> 	// each annotation is set by the optbuilder when the optimizer processes
> 	// the DO block statements.
> 	Annotations tree.Annotations
> ```

Do we also need to save annotations in other top-level blocks? Or in other places that parse PL/pgSQL?
___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 1221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8ZQH-3ASDj9YW-27X:-OY8ZQH-3ASDj9YW-27Y:b-2eovha) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/sem/plpgsqltree/statements.go#L1221)):*
> ```Go
> 	Scroll      tree.CursorScrollOption
> 	Query       tree.Statement
> 	Annotations *tree.Annotations
> ```

Should this be removed?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151849)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/plpgsql/parser/lexer.go` line 188 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8W0lYBSYe4bOfRAh3:-OY9VxDI8jTr-rMI08Qa:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/plpgsql/parser/lexer.go#L188)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

If this has to be done after every call to ParseOneWithOptions, would it make sense to wrap ParseOneWithOptions in a helper function?
</blockquote></details>

Good point, done.
___
*[`pkg/sql/plpgsql/parser/parse.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8ZU5g0aGXwuY5wf9g:-OY9WzEU5QPsUNsnhXev:b4twcec) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/plpgsql/parser/parse.go#L137)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

What will happen to other callers of Parse? It seems like we will use unique annotation indexes, and will set NumAnnotations, but then not actually save the annotations anywhere? Will that interfere with formatting of a CREATE FUNCTION statement for example?
</blockquote></details>

The other callers of `Parse` shouldn't regress. If there is a spot where we call this method and we should be paying attention to annotations, then before this patch the annotations would reference the same information (that happens to be in scope, probably for other nodes), so we would've resolved to incorrect names; after this patch we'll have unique annotation indexes at least, but they would still map to incorrect names. This change would make it more likely that we'd hit index-out-bounds which is why I included the second commit.

I skimmed through the callsites, and the ones I'm concerned about are 4 spots in `opt/optbuilder` like in `buildRoutine` and `buildCreateFunction`. Some other spots where we use `tree.FmtSerializable` also seem somewhat concerning. Yet we do have tests with UDTs in routines, so things must work (?).

In order to de-risk the backport we could apply the new behavior only to the DO block and leave other callsites intact if you think it's worth it. I'm neutral to this idea.
___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 856 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8Z9yM9OT8ToAiEYSe:-OY9TZP15ayT19KUqtkN:b52iift) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/sem/plpgsqltree/statements.go#L856)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should this be removed?
</blockquote></details>

Totally, nice catch, thanks.
___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 1113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8XSoq5nMNMLhtV4Ij:-OY9ZBGCCrbStjFD0RnU:bgouy1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/sem/plpgsqltree/statements.go#L1113)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Do we also need to save annotations in other top-level blocks? Or in other places that parse PL/pgSQL?
</blockquote></details>

Perhaps "yes" to the latter question - which is why I left a TODO. Let me do some debugging to understand why existing tests with UDTs work.
___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 1221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8ZQH-3ASDj9YW-27X:-OY9Tbgl2VPBRCgDHLoq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/sem/plpgsqltree/statements.go#L1221)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Should this be removed?
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/do` line 305 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY88mei1Hf6aNH23loU:-OY9UA-4BUqN5e698UZ3:blxtdfs) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/logictest/testdata/logic_test/do#L305)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Is there a way to test DO blocks that call other DO blocks?
</blockquote></details>

No, it appears that DO blocks have to be stand-alone top-level "anonymous execution blocks" - I'm getting an error trying to have them nested, and ChatGPT also agrees with that. PG docs aren't super clear about this though.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151849)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 1113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8XSoq5nMNMLhtV4Ij:-OY9jslI9ediqvQ5fXpz:b-xokl6r) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/sem/plpgsqltree/statements.go#L1113)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Perhaps "yes" to the latter question - which is why I left a TODO. Let me do some debugging to understand why existing tests with UDTs work.
</blockquote></details>

It looks like things mostly work, but I did find an existing issue that seems to be related to annotations being ignored in one of the callsites #152233. I'll take a closer look at that tomorrow.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151849)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 1113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8XSoq5nMNMLhtV4Ij:-OY9kYjcAzf_k6agxrjE:blsstkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/sem/plpgsqltree/statements.go#L1113)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

It looks like things mostly work, but I did find an existing issue that seems to be related to annotations being ignored in one of the callsites #152233. I'll take a closer look at that tomorrow.
</blockquote></details>

Oh, I should've been explicit - the issue exists before and after this patch.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> @michae2 reviewed 3 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151849)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 1113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8XSoq5nMNMLhtV4Ij:-OYIw2wb04RBcIyE_pKk:b-731iiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/sem/plpgsqltree/statements.go#L1113)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Oh, I should've been explicit - the issue exists before and after this patch.
</blockquote></details>

Thanks for opening that issue! I have to ask, since we're here: would it be hard to fix that issue as well? (IMO that issue seems more important to fix than #151848, since we want to show CREATE FUNCTION statements in bundles and for other debugging.)
___
*[`pkg/sql/logictest/testdata/logic_test/do` line 305 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY88mei1Hf6aNH23loU:-OYItCup01oxS7j9FNx0:b-nav91e) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/logictest/testdata/logic_test/do#L305)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

No, it appears that DO blocks have to be stand-alone top-level "anonymous execution blocks" - I'm getting an error trying to have them nested, and ChatGPT also agrees with that. PG docs aren't super clear about this though.
</blockquote></details>

Thanks for checking!


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151849)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/sem/plpgsqltree/statements.go` line 1113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8XSoq5nMNMLhtV4Ij:-OYJc9682TQ4YtJswpB4:b2c6gb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/sem/plpgsqltree/statements.go#L1113)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Thanks for opening that issue! I have to ask, since we're here: would it be hard to fix that issue as well? (IMO that issue seems more important to fix than #151848, since we want to show CREATE FUNCTION statements in bundles and for other debugging.)
</blockquote></details>

Turns out that it's a known limitation #87475. The difference with this issue is that we allowed to create the routine in the first place - i.e. it looks like SQL routines are currently disabled at creation time whereas PLpgSQL can get created but can have problems down the line, so I don't think we should block this change on fixing either of those issues.

That said, I still plan to spend more a bit more time on this next week to see whether I can find another problematic reproduction with one of those `Parse` callsites.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151849)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/plpgsql/parser/parse.go` line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-OY8ZU5g0aGXwuY5wf9g:-OYiRLVa0-BilfOvTyk8:b-j3t6e4) ([raw file](https://github.com/cockroachdb/cockroach/blob/594f62eb99e42c1b5d8bf301e626932956d439f4/pkg/sql/plpgsql/parser/parse.go#L137)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The other callers of `Parse` shouldn't regress. If there is a spot where we call this method and we should be paying attention to annotations, then before this patch the annotations would reference the same information (that happens to be in scope, probably for other nodes), so we would've resolved to incorrect names; after this patch we'll have unique annotation indexes at least, but they would still map to incorrect names. This change would make it more likely that we'd hit index-out-bounds which is why I included the second commit.

I skimmed through the callsites, and the ones I'm concerned about are 4 spots in `opt/optbuilder` like in `buildRoutine` and `buildCreateFunction`. Some other spots where we use `tree.FmtSerializable` also seem somewhat concerning. Yet we do have tests with UDTs in routines, so things must work (?).

In order to de-risk the backport we could apply the new behavior only to the DO block and leave other callsites intact if you think it's worth it. I'm neutral to this idea.
</blockquote></details>

I audited the callers a bit more thoroughly and was unable to reproduce any problems. I added a somewhat-hand-wavy comment for why ignoring the number of placeholders is ok in most call sites, so I removed the TODO. I also extended the test coverage in one spot where I noticed we didn't have UDTs.

I don't have enough confidence that everything works around this part of the code base, but at least the second commit in this PR will prevent crashes.

There is also this bit https://github.com/cockroachdb/cockroach/blob/3d5ef9464d282f18aaab842d2dc0bd5d49353f41/pkg/sql/sem/tree/object_name.go#L240

meaning that annotations are really only used when either `FmtAlwaysQualifyTableNames` is used or when we have the table name formatter. In at least half call sites I looked at this doesn't seem to be the case.

It doesn't seem worth spending more time on it at this moment, so I think we should merge the patch as is.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
friendly ping @michae2 @DrewKimball

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for looking into this.

> @DrewKimball reviewed 4 of 11 files at r1, 1 of 3 files at r3, 8 of 8 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-:-OZz53CZ0KXzlR32VEuG:b6z2v8s)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Sorry for the delay!

> @michae2 reviewed 8 of 8 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/151849#-:-O_Tgv21BjlQoQ7zIt5W:b-8yidu2)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->
